raphtory/src/io/arrow/df_loaders.rs

```diff
@@ -108,38 +117,89 @@ pub(crate) fn load_nodes_from_df<
-                graph.internal_add_node(t, node_id, &t_props)?;
-                let c_props: Vec<_> = c_props
-                    .chain(shared_constant_properties.iter().cloned())
-                    .collect();
-                if !c_props.is_empty() {
-                    graph.internal_add_constant_node_properties(node_id, &c_props)?;
+            .zip(node_col_resolved.par_iter_mut())
+            .try_for_each(|(gid, resolved)| {
+                let gid = gid.ok_or(LoadError::FatalError)?;
+                let vid = write_locked_graph
+                    .resolve_node(gid)
+                    .map_err(|_| LoadError::FatalError)?;
+                if let Some(cache) = cache {
+                    cache.resolve_node(vid, gid);
                 }
-                Ok::<(), GraphError>(())
+                *resolved = vid.inner();
+                Ok::<(), LoadError>(())
             })?;
+
+        let g = write_locked_graph.graph;
+        let update_time = |time| g.update_time(time);
+
+        write_locked_graph
+            .nodes
+            .resize(write_locked_graph.num_nodes());
+
+        write_locked_graph
+            .nodes
+            .par_iter_mut()
+            .try_for_each(|mut shard| {
+                let mut t_props = vec![];
+                let mut c_props = vec![];
+
+                for (idx, (((vid, time), node_type), gid)) in node_col_resolved
+                    .iter()
+                    .zip(time_col.iter())
+                    .zip(node_type_col_resolved.iter())
+                    .zip(node_col.iter())
+                    .enumerate()
+                {
+                    let shard_id = shard.shard_id();
+                    let node_exists = if let Some(mut_node) = shard.get_mut(*vid) {
+                        mut_node.init(*vid, gid);
+                        mut_node.node_type = *node_type;
+                        t_props.clear();
+                        t_props.extend(prop_cols.iter_row(idx));
+
+                        c_props.clear();
+                        c_props.extend(const_prop_cols.iter_row(idx));
+                        c_props.extend_from_slice(&shared_constant_properties);
+
+                        if let Some(caches) = cache_shards.as_ref() {
+                            let cache = &caches[shard_id];
+                            cache.add_node_update(
+                                TimeIndexEntry(time, start_id + idx),
+                                *vid,
+                                &t_props,
+                            );
+                            cache.add_node_cprops(*vid, &c_props);
+                        }
+
+                        for (id, prop) in c_props.drain(..) {
+                            mut_node.add_constant_prop(id, prop)?;
+                        }
+
+                        true
+                    } else {
+                        false
+                    };
+
+                    if node_exists {
```

Comment:
This can just go directly where you return true on line 186?

Reply:
The reason it's here is because `shard.get_mut` and `shard.t_prop_log_mut` both borrow shard mutably so.. no